pkg/local-pv-monitor/monitor.go

```diff
@@ -0,0 +1,508 @@
+		}
+		return
+	}
+	// TODO: make sure that PV used bytes is not greater that PV capacity ?
+
+	// unmark PV if it was marked before
+	err = monitor.markOrUnmarkPV(pv, MisMatchedVolSize, "yes", false)
+	if err != nil {
+		glog.Errorf("mark PV: %s failed, err: %v", pv.Name, err)
+	}
+	return
+
+}
+
+func (monitor *LocalPVMonitor) checkPVAndBlockSize(mountPath string, pv *v1.PersistentVolume) {
+	capacityByte, err := monitor.VolUtil.GetBlockCapacityByte(mountPath)
+	if err != nil {
+		glog.Errorf("Path %q block stats error: %v", mountPath, err)
+		return
+	}
+	// actually if PV is provisioned by provisioner, the two values must be equal, but the PV may be
+	// created manually, so the PV capacity must not be greater than block device capacity
+	storage := pv.Spec.Capacity[v1.ResourceStorage]
+	if util.RoundDownCapacityPretty(capacityByte) < storage.Value() {
+		// mark PV and send a event
+		err = monitor.markOrUnmarkPV(pv, MisMatchedVolSize, "yes", true)
+		if err != nil {
+			glog.Errorf("mark PV: %s failed, err: %v", pv.Name, err)
+		}
+		return
+	}
+	// TODO: make sure that PV used bytes is not greater that PV capacity ?
+
+	// unmark PV if it was marked before
+	err = monitor.markOrUnmarkPV(pv, MisMatchedVolSize, "yes", false)
+	if err != nil {
+		glog.Errorf("mark PV: %s failed, err: %v", pv.Name, err)
+	}
+	return
+}
+
+// CheckNodeAffinity looks at the PV node affinity, and checks if the node has the same corresponding labels
+// This ensures that we don't mount a volume that doesn't belong to this node
+func CheckNodeAffinity(pv *v1.PersistentVolume, nodeLabels map[string]string) (bool, error) {
+	affinity, err := helper.GetStorageNodeAffinityFromAnnotation(pv.Annotations)
+	if err != nil {
+		return false, fmt.Errorf("error getting storage node affinity: %v", err)
+	}
+	if affinity == nil {
+		return false, nil
+	}
+
+	if affinity.RequiredDuringSchedulingIgnoredDuringExecution != nil {
+		terms := affinity.RequiredDuringSchedulingIgnoredDuringExecution.NodeSelectorTerms
+		glog.V(10).Infof("Match for RequiredDuringSchedulingIgnoredDuringExecution node selector terms %+v", terms)
+		for _, term := range terms {
+			selector, err := helper.NodeSelectorRequirementsAsSelector(term.MatchExpressions)
+			if err != nil {
+				return false, fmt.Errorf("failed to parse MatchExpressions: %v", err)
+			}
+			if !selector.Matches(labels.Set(nodeLabels)) {
+				return false, fmt.Errorf("NodeSelectorTerm %+v does not match node labels", term.MatchExpressions)
+			}
+		}
+	}
+	return true, nil
+}
+
+// markPV marks PV by adding annotation
+func (monitor *LocalPVMonitor) markOrUnmarkPV(pv *v1.PersistentVolume, ann, value string, mark bool) error {
+	// The volume from method args can be pointing to watcher cache. We must not
+	// modify these, therefore create a copy.
+	volumeClone := pv.DeepCopy()
+	var eventMes string
+
+	if mark {
+		// mark PV
+		_, ok := volumeClone.ObjectMeta.Annotations[ann]
```

Comment:
so right now, then only thing we do is to to mark the PV using annotation and send an event right?

Reply:
yes, since we do not have taint API, so use annotations instead